torch/csrc/lazy/core/hash.cpp

```diff
@@ -0,0 +1,81 @@
+#include <iomanip>
+#include <sstream>
+
+#include <torch/csrc/lazy/core/hash.h>
+
+namespace torch {
+namespace lazy {
+namespace {
+
+hash_t LoadHash(const uint8_t** data, const uint8_t* top) {
+  std::ptrdiff_t size = top - (*data);
+  if (size >= (int)sizeof(hash_t)) {
+    hash_t v;
+    std::memcpy(&v, *data, sizeof(v));
+    *data += sizeof(hash_t);
+    return v;
+  }
+  union {
+    hash_t h;
+    std::array<uint8_t, sizeof(hash_t)> b;
+  } uval = {.h = 0};
+  std::memcpy(uval.b.data(), *data, size);
```

Comment:
I think this part is from pt/xla, but I am curious if you know why do we need to memcpy to `uval.b.data()` here?

Reply:
I don't see the necessity of doing it this way.  Clearly the aim is to copy a 128-bit chunk from an arbitrary point in a uint8_t buffer, and then return it as a hash_t.

But why bother with a union and an explicit memcpy?  Couldn't we cast *data to a hash_t* and use that to initialize a hash_t for return?  (And not need the union type at all?)

It's possible someone did it this way to work around some kind of linter requirement.

Reply:
The `memcpy()` is required to deal with architectures which do not allow unaligned memory access (or they pay big penalty).
Note that these days a `memcpy()` maps to a compiler builtin, and the compiler uses the most appropriate code depending on the architecture.
Example, on x86_64 this is just a move:

https://godbolt.org/z/4escG8WWE

The union is needed to read the last partial bytes, padding the missing ones with zeros.